website/versioned_docs/version-2.0.0-beta.6/azure-functions.md

```diff
@@ -0,0 +1,116 @@
+---
+title: Azure Functions Integration
+---
+
+## Using TypeGraphQL in Microsoft Azure Functions
+
+Integrating TypeGraphQL with Azure Functions involves the following key steps:
+
+1. Generate GraphQL schema based on your resolvers
+2. Notify Apollo Server about your schema
+
+Below is how you can implement the azure function entry point (with explainations in-line):
+
+```ts
+// index.ts
+
+import 'reflect-metadata';
+import path from 'path';
+import { ApolloServer } from '@apollo/server';
+import { startServerAndCreateHandler } from '@as-integrations/azure-functions';
+import { buildSchemaSync } from 'type-graphql';
+import { Container } from 'typedi';
+import { GraphQLFormattedError } from 'graphql';
+import { UserResolver } from 'YOUR_IMPORT_PATH'; // TypeGraphQL Resolver
+import { AccountResolver } from 'YOUR_IMPORT_PATH';  // TypeGraphQL Resolver
+
+// Bundle resolvers to build the schema
+const schema = buildSchemaSync({
```

Comment:
Is this necessary to build the schema sync? It's not recommended approach for TypeGraphQL

Reply:
Yes it is.

In apollo-server, the `schema` field is expecting `GraphQLSchema`, not the one wrapped in a promise.

```ts
import { ApolloServer } from '@apollo/server';

// using buildSchemaSync
const server = new ApolloServer({
  schema, // type: GraphQLSchema ✅
  introspection: true,
});

// using buildSchema
const server = new ApolloServer({
  schema, // type: Promise<GraphQLSchema> ❌
  introspection: true,
});
```

When I used `buildSchema`, I got this error:
```text
Type Promise<GraphQLSchema> is missing the following properties from type:
description
extensions
astNode
extensionASTNodes
and 20 more.
```

I see on type-graphql doc that it uses `buildSchema` for examples integrating with apollo-server. If I missed something, let me know.

Reply:
Yeah, you have to read about `await` and the top-level await support.
I bet Azure handlers can work with async functions too.

Reply:
I see why `buildSchema` can be beneficial in some cases.

But we've just tried doing this:
```ts
// indes.ts

let server: ApolloServer;

const startServer = async () => {
  const schema = await buildSchema({
    resolvers: [
      EmployeeResolver,
      HistoryEmployeeResolver,
      CurrentStateRoleResolver,
      HistoryCurrentStateRoleResolver,
      FutureStateRoleResolver,
      HistoryFutureStateRoleResolver,
      ChangeProcessResolver,
      ClearinghouseResolver,
    ],

    emitSchemaFile:
      env.NODE_ENV === 'local' ? path.resolve('./src/schema.graphql') : false,
    container: Container,
    validate: true,
  });

  server = new ApolloServer({
    schema,
    introspection: true,
    formatError: (err: GraphQLFormattedError) => err,
  });
};

startServer().catch(console.error);

export default startServerAndCreateHandler(server, {
  context: async (args: AzureFunctionsContextFunctionArgument) => {
    return await transformContext(args);
  },
});
```
and we got this error: `Worker was unable to load function handler-graphql: 'Cannot read properties of undefined (reading 'startInBackgroundHandlingStartupErrorsByLoggingAndFailingAllRequests')'
[2024-03-10T21:37:31.043Z`

Seems like `startServerAndCreateHandler()` from `'@as-integrations/azure-functions` is unable to load schemas wrapped in a promise, I could be wrong.

Still, I won't see why we need to stick to using `buildSchema` just because our resolver methods use asyncronous operations. Are you able to setup an azure function without using `buildSchemaSync`?

Reply:
> Are you able to setup an azure function without using buildSchemaSync?

I don't use Azure but it's easily possible for AWS. Maybe you can use the same pattern.

> Still, I won't see why we need to stick to using buildSchema just because our resolver methods use asyncronous operations. Don't we all need to first build a proper gql schema before starting the server?

Sync build schema is just a workaround for envs that can't handle the promise version. Sync is always imitating and might not have all the stuff available in async version, so at least docs should show how to use them properly.